internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,385 @@
+		jobHandle,
+		iocpChan,
+		sync.RWMutex{},
+	}, nil
+}
+
+// NotificationChan returns a channel that can be used to receive notifications surrounding
+// the job object. The caller will have to type assert the type of message received.
+func (job *JobObject) NotificationChan() chan interface{} {
+	return job.notifChan
+}
+
+// Close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *JobObject) Close() error {
+	if job.jobHandle == 0 {
+		return nil
+	}
+	job.handleLock.Lock()
+	defer job.handleLock.Unlock()
+
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			return err
+		}
+		// Handles now invalid so if the map entry to receive notifications for this job still
+		// exists remove it so we can stop receiving notifications.
+		if _, ok := jobMap[uintptr(job.jobHandle)]; ok {
+			delete(jobMap, uintptr(job.jobHandle))
+		}
+		job.jobHandle = 0
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *JobObject) Assign(pid uint32) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if pid == 0 {
+		return errors.New("process has not started")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, pid)
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminate terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *JobObject) Terminate() error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+	return windows.TerminateJobObject(job.jobHandle, 1)
+}
+
+// Shutdown gracefully shuts down all the processes in the job object.
+func (job *JobObject) Shutdown(ctx context.Context) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	var (
+		terminate bool
+		signalErr bool
+	)
+	pids, err := job.Pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Errorf("failed to send ctrl-break to process in job: %s", err)
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again to check if they were successfully killed by the signals.
+	// Certain programs handle ctrl signals so we need a fallback to actually kill them.
+	newPids, err := job.Pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.Terminate()
```

Comment:
This seems like we send a control event to allow graceful termination, then don't give the process any time to actually shutdown gracefully, and immediately terminate if the process has not exited. Most processes will need some time to shutdown gracefully as this implies things like flushing cache, closing connections, etc. I would suggest instead sending the control events, and then using `ctx` to determine how long to wait, and then doing a terminate at that point.